File: sorting-searching/sorting-algorithms/replacement-selection-sort/replacement-selection-sort.py
import heapq

def merge_two_sorted_runs(left, right):
    result = []

    i = 0
    j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    while i < len(left):
        result.append(left[i])
        i += 1

    while j < len(right):
        result.append(right[j])
        j += 1

    return result


def merge_all_runs(runs):
    if not runs:
        return []

    result = runs[0]

    for i in range(1, len(runs)):
        result = merge_two_sorted_runs(result, runs[i])

    return result


def generate_runs(input_array, memory_size):
    runs = []
    current_heap = []
    frozen = []

    index = 0

    while index < len(input_array) and len(current_heap) < memory_size:
        heapq.heappush(current_heap, input_array[index])
        index += 1

    current_run = []
    last_output = float("-inf")

    while current_heap:
        smallest = heapq.heappop(current_heap)

        current_run.append(smallest)
        last_output = smallest

        if index < len(input_array):
            next_value = input_array[index]
            index += 1

            if next_value >= last_output:
                heapq.heappush(current_heap, next_value)
            else:
                frozen.append(next_value)

        if not current_heap:
            runs.append(current_run)
            current_run = []

            for value in frozen:
                heapq.heappush(current_heap, value)

            frozen = []
            last_output = float("-inf")

    return runs


def replacement_selection_sort(input_array, memory_size):
    runs = generate_runs(input_array, memory_size)
    return merge_all_runs(runs)


arr = [6, 2, 8, 3, 9, 1, 7, 4, 5]
memory_size = 3

runs = generate_runs(arr, memory_size)

print("Generated Runs:")
for i, run in enumerate(runs, start=1):
    print(f"Run {i}: {run}")

sorted_array = merge_all_runs(runs)

print("Final Sorted Output:", sorted_array)

File: sorting-searching/sorting-algorithms/replacement-selection-sort/ReplacementSelectionSort.java
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;

public class ReplacementSelectionSort {

    static List<Integer> mergeTwoSortedRuns(List<Integer> left, List<Integer> right) {
        List<Integer> result = new ArrayList<>();

        int i = 0;
        int j = 0;

        while (i < left.size() && j < right.size()) {
            if (left.get(i) <= right.get(j)) {
                result.add(left.get(i));
                i++;
            } else {
                result.add(right.get(j));
                j++;
            }
        }

        while (i < left.size()) {
            result.add(left.get(i));
            i++;
        }

        while (j < right.size()) {
            result.add(right.get(j));
            j++;
        }

        return result;
    }

    static List<Integer> mergeAllRuns(List<List<Integer>> runs) {
        if (runs.isEmpty()) {
            return new ArrayList<>();
        }

        List<Integer> result = runs.get(0);

        for (int i = 1; i < runs.size(); i++) {
            result = mergeTwoSortedRuns(result, runs.get(i));
        }

        return result;
    }

    static List<List<Integer>> generateRuns(int[] inputArray, int memorySize) {
        List<List<Integer>> runs = new ArrayList<>();
        PriorityQueue<Integer> currentHeap = new PriorityQueue<>();
        List<Integer> frozen = new ArrayList<>();

        int index = 0;

        while (index < inputArray.length && currentHeap.size() < memorySize) {
            currentHeap.add(inputArray[index]);
            index++;
        }

        List<Integer> currentRun = new ArrayList<>();
        int lastOutput = Integer.MIN_VALUE;

        while (!currentHeap.isEmpty()) {
            int smallest = currentHeap.poll();

            currentRun.add(smallest);
            lastOutput = smallest;

            if (index < inputArray.length) {
                int nextValue = inputArray[index];
                index++;

                if (nextValue >= lastOutput) {
                    currentHeap.add(nextValue);
                } else {
                    frozen.add(nextValue);
                }
            }

            if (currentHeap.isEmpty()) {
                runs.add(currentRun);
                currentRun = new ArrayList<>();

                for (int value : frozen) {
                    currentHeap.add(value);
                }

                frozen.clear();
                lastOutput = Integer.MIN_VALUE;
            }
        }

        return runs;
    }

    static List<Integer> replacementSelectionSort(int[] inputArray, int memorySize) {
        List<List<Integer>> runs = generateRuns(inputArray, memorySize);
        return mergeAllRuns(runs);
    }

    public static void main(String[] args) {
        int[] arr = {6, 2, 8, 3, 9, 1, 7, 4, 5};
        int memorySize = 3;

        List<List<Integer>> runs = generateRuns(arr, memorySize);

        System.out.println("Generated Runs:");
        for (int i = 0; i < runs.size(); i++) {
            System.out.println("Run " + (i + 1) + ": " + runs.get(i));
        }

        List<Integer> sortedArray = mergeAllRuns(runs);

        System.out.println("Final Sorted Output: " + sortedArray);
    }
}
